eessi/reframe/eessi_utils/hooks.py

```diff
@@ -15,109 +17,167 @@
+    if test.max_avail_cpus_per_node is None:
         raise AttributeError(PROCESSOR_INFO_MISSING)
 
-    if compute_unit == 'gpu':
+    # Check if either node_part, or default_num_cpus_per_node and default_num_gpus_per_node are set correctly
+    if not (
+        type(test.node_part) == int or
+        (type(test.default_num_cpus_per_node) == int and type(test.default_num_gpus_per_node) == int)
+    ):
+        raise ValueError(
+            f'Either node_part ({test.node_part}), or default_num_cpus_per_node ({test.default_num_cpus_per_node}) and'
+            f' default num_gpus_per_node ({test.default_num_gpus_per_node}) must be defined and have integer values.'
+        )
+
+    # Check if the default number of cpus per node is already defined in the test
+    # (e.g. by earlier hooks like set_tag_scale).
+    # If so, check if it doesn't exceed the maximum available.
+    # If not, set default_num_cpus_per_node based on the maximum available cpus and node_part
+    if test.default_num_cpus_per_node:
+        # may skip if not enough CPUs
+        test.skip_if(
+            test.default_num_cpus_per_node > test.max_avail_cpus_per_node,
+            f'CPUs per node ({test.default_num_cpus_per_node}) is higher than max available'
+            f' ({test.max_avail_cpus_per_node}) in current partition ({test.current_partition.name})'
+        )
+    else:
+        # no default set yet, so setting one
+        test.default_num_cpus_per_node = int(test.max_avail_cpus_per_node / test.node_part)
+
+    if compute_unit == DEVICES['GPU']:
         assign_one_task_per_gpu(test)
-    elif compute_unit == 'cpu':
+    elif compute_unit == DEVICES['CPU']:
         assign_one_task_per_cpu(test)
     else:
         raise ValueError(f'compute unit {compute_unit} is currently not supported')
 
 
-def assign_one_task_per_cpu(test: rfm.RegressionTest) -> rfm.RegressionTest:
+def assign_one_task_per_cpu(test: rfm.RegressionTest):
     """
     Sets num_tasks_per_node and num_cpus_per_task such that it will run one task per core,
-    unless specified with --setvar num_tasks_per_node=<x> and/or --setvar num_cpus_per_task=<y>.
+    unless specified with:
+    --setvar num_tasks_per_node=<x> and/or
+    --setvar num_cpus_per_task=<y>.
+
+    Variables:
+    - default_num_cpus_per_node: default number of CPUs per node as defined in the test
+    (e.g. by earlier hooks like set_tag_scale)
+
+
+    Default resources requested:
+    - num_tasks_per_node = default_num_cpus_per_node
+    - num_cpus_per_task = default_num_cpus_per_node / num_tasks_per_node
     """
-    max_cpus_per_node = test.max_cpus_per_node
-    num_tasks_per_node = test.num_tasks_per_node
-    num_cpus_per_task = test.num_cpus_per_task
+    # neither num_tasks_per_node nor num_cpus_per_node are set
+    if not test.num_tasks_per_node and not test.num_cpus_per_task:
+        test.num_tasks_per_node = test.default_num_cpus_per_node
```

Comment:
This `if` condition should _also_ set `test.num_cpus_per_task =1` 

Or, according to the description of the function, it should set `test.num_cpus_per_task = test.default_num_cpus_per_node / test.num_tasks_per_node`, but... that's `1` since we just defined `test.num_tasks_per_node = test.default_num_cpus_per_node`. So we might as well set it to `1` right away. (and _that_ is also in the description of `assign_one_task_per_compute_unit` by the way)

Reply:
good catch, fixed in https://github.com/EESSI/test-suite/pull/28/commits/92db99ed21e7e0ed216dc911907de1ae27de0e01